src/DynamoCoreWpf/Controls/InfiniteGridView.xaml.cs

```diff
@@ -0,0 +1,235 @@
+            var workspaceView = WpfUtilities.FindUpVisualTree<WorkspaceView>(this);
+
+            if (workspaceView == null)
+            {
+                throw new InvalidOperationException(
+                    "InfiniteGridView should be a nested element of WorkspaceView");
+            }
+
+            workspaceModel = workspaceView.ViewModel.Model;
+        }
+
+        private void UpdateDrawingVisual()
+        {
+            if (workspaceModel == null)
+                return;
+
+            UpdateDrawingVisual(workspaceModel.X, workspaceModel.Y, workspaceModel.Zoom);
+        }
+
+        private void UpdateDrawingVisual(double x, double y, double zoom)
+        {
+            #region Scale Adjustment
+
+            // Bring scale factor to within zoom boundaries.
+            var localScale = zoom;
+            while (localScale * MajorGridLineSpacing < MinMajorGridSpacing)
+                localScale = localScale * ScaleFactor;
+            while (localScale * MajorGridLineSpacing > MaxMajorGridSpacing)
+                localScale = localScale / ScaleFactor;
+
+            #endregion
+
+            #region Positional Adjustment
+
+            // The scale is know, adjust the top-left corner.
+            var startX = x;
+            var startY = y;
+            var scaledMajorGridSpacing = localScale * MajorGridLineSpacing;
+
+            if (startX > 0)
+            {
+                // Assuming after applying scale factor, the major grid lines 
+                // are 10px apart. If the current WorkspaceModel.X is 24px, then 
+                // 
+                //      v = floor(24/10) = floor(2.4) = 2
+                //      w = 24 - 2 x 10 = 4
+                // 
+                // We could start drawing the major grid line from 4px, but that 
+                // leaves a gap at the left edge. So it would be nice if major 
+                // grid line starts from 4 - 10 = -6px, so that it appears that 
+                // the left-most grid line is beyond the left edge.
+                // 
+                var v = ((int)Math.Floor(startX / scaledMajorGridSpacing));
+                var w = (startX - (v * scaledMajorGridSpacing));
+                startX = w - scaledMajorGridSpacing;
+            }
+            else if (startX < -scaledMajorGridSpacing)
+            {
+                // Assuming after applying scale factor, the major grid lines 
+                // are 10px apart. If the current WorkspaceModel.X is -24px, then 
+                // 
+                //      startX = abs(-24) = 24
+                //      v = floor(24/10) = floor(2.4) = 2
+                //      w = 24 - 2 x 10 = 4
+                //      startX = -w = -4
+                // 
+                startX = Math.Abs(startX);
+                var v = ((int)Math.Floor(startX / scaledMajorGridSpacing));
+                var w = (startX - (v * scaledMajorGridSpacing));
+                startX = -w;
+            }
+
+            if (startY > 0)
```

Comment:
Oh, i'm not saying the while loop below, but just the computation of `startX` and `startY`, seems the computation is the same.